engiopt/vqgan/vqgan.py

```diff
@@ -0,0 +1,1900 @@
+
+        self.apply(self._init_weights)
+        for pn, p in self.named_parameters():
+            if pn.endswith("c_proj.weight"):
+                th.nn.init.normal_(p, mean=0.0, std=0.02 / math.sqrt(2 * config.n_layer))
+
+    def _init_weights(self, module: nn.Module) -> None:
+        if isinstance(module, nn.Linear):
+            th.nn.init.normal_(module.weight, mean=0.0, std=0.02)
+            if module.bias is not None:
+                th.nn.init.zeros_(module.bias)
+        elif isinstance(module, nn.Embedding):
+            th.nn.init.normal_(module.weight, mean=0.0, std=0.02)
+
+    def forward(
+        self,
+        idx: th.Tensor,
+        targets: th.Tensor | None = None,
+    ) -> tuple[th.Tensor, th.Tensor | None]:
+        """Forward pass returning logits and optional cross-entropy loss."""
+        device = idx.device
+        _, t = idx.size()
+        assert t <= self.config.block_size, f"Cannot forward sequence of length {t}; block size is {self.config.block_size}"
+        pos = th.arange(0, t, dtype=th.long, device=device)
+
+        tok_emb = self.transformer["wte"](idx)
+        pos_emb = self.transformer["wpe"](pos)
+        x = self.transformer["drop"](tok_emb + pos_emb)
+        for block in self.transformer["h"]:
+            x = block(x)
+        x = self.transformer["ln_f"](x)
+
+        logits = self.lm_head(x)
+        loss: th.Tensor | None
+        if targets is not None:
+            loss = f.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1), ignore_index=-1)
+        else:
+            loss = None
+        return logits, loss
+
+
+###########################################
+########## GPT-2 BASE CODE ABOVE ##########
+###########################################
+
+
+class VQGANTransformer(nn.Module):
+    """Wrapper for VQGAN Stage 2: Transformer.
+
+    Generative component of VQGAN trained on the Stage 1 discrete latent space.
+
+    Parameters:
+        conditional (bool): If True, use CVQGAN for conditioning.
+        vqgan (VQGAN): Pretrained VQGAN model for primary image encoding/decoding.
+        cvqgan (VQGAN): Pretrained CVQGAN model for conditional encoding (if conditional=True).
+        image_size (int): Input image size (assumed square).
+        decoder_channels (tuple[int, ...]): Decoder channels from the VQGAN model.
+        cond_feature_map_dim (int): Feature map dimension from the CVQGAN encoder (if conditional=True).
+        num_codebook_vectors (int): Number of codebook vectors from the VQGAN model.
+        n_layer (int): Number of Transformer layers.
+        n_head (int): Number of attention heads in the Transformer.
+        n_embd (int): Embedding dimension in the Transformer.
+        dropout (float): Dropout rate in the Transformer.
+        bias (bool): If True, use bias terms in the Transformer layers.
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        *,
+        conditional: bool = True,
```

Comment:
Is this correct python? I thought you can only put arguments with default values after arguments without default